fmaas-router/src/rpc/info.rs

```diff
@@ -0,0 +1,76 @@
+use std::collections::HashMap;
+
+use ginepro::LoadBalancedChannel;
+use tonic::{transport::ClientTlsConfig, Request, Response, Status};
+use tracing::{debug, instrument};
+
+use crate::rpc::extract_model_id;
+
+use crate::{create_clients, pb::{
+    caikit::runtime::info::{
+        info_service_client::InfoServiceClient, info_service_server::InfoService
+    },
+    caikit_data_model::common::runtime::{
+        ModelInfoRequest, ModelInfoResponse, RuntimeInfoRequest, RuntimeInfoResponse
+    }
+}, ServiceAddr};
+
+
+#[derive(Debug, Default)]
+pub struct InfoServicer {
+    clients: HashMap<String, InfoServiceClient<LoadBalancedChannel>>,
+}
+
+impl InfoServicer {
+    pub async fn new(
+        default_target_port: u16,
+        client_tls: Option<&ClientTlsConfig>,
+        model_map: &HashMap<String, ServiceAddr>,
+    ) -> Self {
+        let clients = create_clients(
+            default_target_port, client_tls, model_map, InfoServiceClient::new
+        ).await;
+        Self { clients }
+    }
+
+    async fn client(
+        &self,
+        model_id: &str,
+    ) -> Result<InfoServiceClient<LoadBalancedChannel>, Status> {
+        Ok(self
+            .clients
+            .get(model_id)
+            .ok_or_else(|| Status::not_found(format!("Unrecognized model_id: {model_id}")))?
+            .clone())
+    }
+}
+
+#[tonic::async_trait]
+impl InfoService for InfoServicer {
+#[instrument(skip_all)]
+async fn get_models_info(
+    &self,
+    request: Request<ModelInfoRequest>,
+) -> Result<Response<ModelInfoResponse>, Status> {
+    let model_id = extract_model_id(&request)?;
+    let mir: &ModelInfoRequest = request.get_ref();
+    if mir.model_ids.is_empty() {
+        return Ok(Response::new(ModelInfoResponse::default()));
+    }
```

Comment:
A bit confused on the logic here, it looks like the `ModelInfoRequest::model_ids` field contains a list of model ids, but if they're sent we're then instead grabbing a model id from the metadata and returning info for that?

Reply:
Hi! It takes a list of model ids, and returns a list of metadata dicts for all of them - maybe there is a better way to do this? Or clearer?

Reply:
huh, looking back I may have put this comment on the wrong line lol

If I have this right, then (roughly):
- The grpc metadata must contain a model id, and this model id must be in the `model_map` that maps model ids to service names
- A client is created to the service that this metadata model id maps to
- The request must also include a non-empty list of model ids
- Then the request is forwarded along to the client we created

I think I kinda get what is going on- we're using the metadata to route to the correct instance of a caikit runtime, and then forwarding along the request with potentially multiple model ids inside of it, and we're making a big assumption that those model ids are models that that caikit runtime knows about.

It's... a little confusing I think. What is a user supposed to put in the model id list? Can one instance of a caikit embeddings runtime have multiple models loaded? Are all of those models in the `model_map`?

I think the big problem here is that the router is supposed to know which models are loaded where, not the caller. So I could see a caller sticking a bunch of model ids in the request, and ending up with a bunch of errors because those models were all loaded on different backend services.

Should the router instead look up where each model in `mir.model_ids` exists, and fan-out requests to each runtime that they're loaded in?

Reply:
@joerunde, I think I understand what happened here, the [user request](https://github.ibm.com/NGP-TWC/ml-planning/issues/36701) was to add "ModelInfo and Tokenize APIs from the GenerationService (generation.proto)". However, since the embeddings service doesn't have our tgis grpc API but is a caikit runtime, the corresponding caikit interfaces were added. I think that, for one thing, it shows that the users intention here is not to query several models.
  
The other point is that none of the other implemented caikit APIs require a list of model IDs. ModelInfoRequest seems to be an exception. So to be pragmatic, my suggestion would be something like this:

```
    async fn get_models_info(
        &self,
        request: Request<ModelInfoRequest>,
    ) -> Result<Response<ModelInfoResponse>, Status> {
        let model_id = extract_model_id(&request)?;
        let mut mir: ModelInfoRequest = request.get_ref().to_owned();
        if mir.model_ids.is_empty() {
            mir.model_ids.push(model_id.to_string());
        }
        if mir.model_ids.len() > 1 || model_id != mir.model_ids[0] {
            // This model info endpoint is being added so that
            // users of the embedding service can obtain more information
            // about that particular service.
            // Calls to model info should follow the same pattern of other calls
            // like TokenizationTaskRequest where the destination model
            // is uniquely specified by the mm-model-id metadata field.
            return Err(Status::unimplemented(
                "Architectural assumption violated: model_ids must be empty or\
                 have the same id as mm-model-id metadata as only element"
            ))
        }
        debug!(
            "Routing get models info request for Model ID {}",
            model_id
        );
        self.client(model_id)
            .await?
            .get_models_info(request)
            .await
    }
```

In other words, we enforce the architectural assumptions and document them, so that if users send a request that violates those assumptions there will be a clear message error message that can be used to find an explanation for the error.

Reply:
@maxdebayser @flaviabeo  I disagree here, it sounds like a pretty bad experience to see an api that accepts a list of model ids and then have that fail when you supply any. It would also make the behavior very different when you access this api via a caikit runtime directly vs. via the router, because:
- caikit does not accept `mm-model-id` for this rpc, and
- caikit has no limitation on how many / which model ids you can put in the request

I really don't want to create a different api here and then have to field questions from watsonx product team about why they have to change how their clients are written, or why the api doesn't work intuitively

Re: the `mm-model-id` metadata, that's in place for inference requests for two main reasons:
- to line up with kserve, so that you can use the multi-model serving mode. (From before we added support to grab a model id from a specific field in the request)
- to line up with how model's inference methods are written. They generally only take inputs and not a model id, and we were trying to make caikit's remote api line up with the python api as much as possible

For the non-inference runtime APIs like this one, caikit was intentionally designed not to use the `mm-model-id` metadata, and instead we put the model id(s) in the request.

I don't think it'd be too hard to put the routing logic in here:
(not-rust pseudocode because I'm only a little rustacean)
```
for model_id in mir.model_ids {
    client = self.client(model_id).await?
    responses.push(client.get_models_info.(request).await?
}

build_final_response(responses)
```

That keeps the same api between the router and caikit, and keeps the router transparently handling all the routing for the caller

Reply:
Nice, this looks great!